kernel/acpi/hpet/src/lib.rs

```diff
@@ -106,6 +107,13 @@ impl Hpet {
     }
 }
 
+impl time::ClockSource for Hpet {
+    type ClockType = time::Monotonic;
+
+    fn now() -> Instant {
+        Instant::new(get_hpet().expect("couldn't get HPET").get_counter())
+    }
```

Comment:
I understand why it works like this, but I still find this API supremely weird. Typically when you impl a trait for a type, you use that type's members in some way.

Obviously for the `now()` fn (and that function only), you'll call it without any args/context, so this is a bit more sensical. However, I would argue that there should be an "inner now" fn that _does_ use the clock type's fields/members along with the rest of the `Clocksource` methods, and then an "outer now" fn that can depend on those.

In any case we can keep this as is for now, but I think the entire `time` API needs to change such that you implement `Clocksource` for specific clock types in a typical way that actually uses their fields & member functions. Then you can have a standalone, contextless/argumentless `now()` adapter around that.

Reply:
Follow-up: the "outer now" fn would basically just hold a reference to the current "best" impl for each clocktype. This is more in line with the idea of registering all clocksources and then ranking them as they are registered. 

The key design concept is that there would never exist a way to invoke `now()` on an `Hpet` clocksource if the `Hpet` did not exist and hadn't been registered; this is how the other device registration APIs work. 

With the current impl, I could invoke `now()` on an `Hpet` device directly and cause a panic even without obtaining a reference to an initialized Hpet device first; that's really what I find strange about this API
```rust
fn main() {
    use time::Clocksource;
    let now = Hpet::now(); // panic
}
```
I realize it wouldn't be called if you go through the `time` API (since it wouldn't've been registered), but it's still pretty odd.

Reply:
We want to minimise the number of atomic function pointers we use. If `Clocksource::now` has a `&self` parameter, then we also need a `Clocksource::get_ref` function that returns a type that dereferences into a `&self`. But then you're back to where you started because the `get_ref` function would have to assume that the static is initialised.

We can't pass a `&'static impl Lockable<'static, dyn ClockSource>` into `time::register_clock_source` because we can't cast a `&'static RwLock<BorrowedMappedPages<Hpet, Mutable>>` into a trait object as the layout of `Hpet` is important.

At some point, the `Clocksource` trait needs to interact with the static, but we can never guarantee that the `static` is initialised. I don't see a way around at least one of the `ClockSource` functions panicking, and although using `now` as we do is confusing, it minimises the number of atomic function pointers we use.